Create NonADFunctorInterface and ADFunctorInterface

lindsayad opened the pull request:
Closes #23311

lindsayad committed:
Create NonADFunctorInterface

Closes #23311

GiudGiud commented on framework/src/auxkernels/TimeDerivativeAux.C:
```diff
@@ -26,11 +24,10 @@ TimeDerivativeAuxTempl<is_ad>::validParams()
-    _functor(getFunctor<GenericReal<is_ad>>("functor")),
-    _factor(getFunctor<GenericReal<is_ad>>("factor")),
+    _functor(getFunctor<Real>("functor")),
+    _factor(getFunctor<Real>("factor")),
```
getting the right one by default, perfect!!

lindsayad committed:
Create ADFunctorInterface

Refs #23311

GiudGiud commented on framework/include/utils/ADifyFunctor.h:
```diff
@@ -0,0 +1,119 @@
+  using typename Moose::FunctorBase<T>::GradientType;
+  using typename Moose::FunctorBase<T>::DotType;
+
+  ADifyFunctor(const FunctorBase<typename MetaPhysicL::RawType<T>::value_type> & non_ad_functor)
```
you used to be against that vehemently?

lindsayad replied:
As you get older you get softer? I don't know what to say haha. Maybe I'm amenable because no one has shot themselves in the foot yet...and maybe after a bunch of people shoot themselves in the foot then I'll be back to my old thought patterns? I do think that the general knowledge that the aux system drops derivatives and that the material system keeps derivatives has percolated more through the user base. Maybe that's my justification

lindsayad replied:
I guess as things stand now users can already shoot themselves in the foot since aux variables are AD functors

GiudGiud replied:
Maybe the immediate INL user base has understood that but it s come up in Discussions. People trying to use auxvariables to get going with some variable conversions from vector to components in NS for example

lindsayad replied:
In the SubProblem I check to see whether someone has requested a non-AD functor in an AD object when the "true" functor is AD. If so, then I error. If someone requests a non-AD functor in a non-AD object when the "true" functor is AD, then the RawValueFunctor is used

GiudGiud commented on framework/include/interfaces/ADFunctorInterface.h:
```diff
@@ -0,0 +1,92 @@
+  const Moose::Functor<T> & getFunctor(const std::string & name, SubProblem & subproblem);
+
+private:
+  /// Storage vector for raw functors
```
```suggestion
  /// Storage vector for AD-extended functors
```

GiudGiud commented on framework/include/interfaces/ADFunctorInterface.h:
```diff
@@ -0,0 +1,92 @@
+            functor_name, _fi_tid, _fi_name);
+    _adified_functors.emplace_back(std::make_unique<Moose::Functor<T>>(
+        std::make_unique<Moose::ADifyFunctor<T>>(non_ad_functor)));
+    return static_cast<Moose::Functor<T> &>(*_adified_functors.back());
```
keep track of all the names, then output an info message ?

lindsayad replied:
This will be undergoing some changes

GiudGiud commented on framework/include/utils/RawValueFunctor.h:
```diff
@@ -0,0 +1,116 @@
+  /**
+   * Forward calls to wrapped object
+   */
+  ValueType evaluate(const ElemArg & elem, unsigned int state = 0) const override
```
could we template all this

lindsayad replied:
What do you mean? The class is templated

lindsayad replied:
template methods cannot be virtual

GiudGiud replied:
template on the functor arguments I meant

GiudGiud replied:
ugh C++

lindsayad replied:
I don't think it's fair to critique C++ because it has two neat features that can't be intermingled 😄

lindsayad replied:
I could create a template method that all the virtual overrides call...but the virtual overrides would go from being one line to one line

GiudGiud replied:
yeah not worth it

GiudGiud reviewed (commented):
some ideas

lindsayad committed:
Smoothly handle appropriate no-AD/AD functor mixing

lindsayad committed:
Remove now extraneous templating on Function

lindsayad committed:
Update input files

lindsayad committed:
Provide backwards compatibility for ADParsedFunction

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1333257/) on 16234d5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23328/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1333293/) on 16234d5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.70%</td>
      <td>84.65%</td>
      <td>-0.05%</td>
      <td>71.07%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>82848</td>
      <td>82910</td>
      <td>+62</td>
      <td>226</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14960</td>
      <td>15031</td>
      <td>+71</td>
      <td>92</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.37%</td>
      <td>85.38%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6671</td>
      <td>6674</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1143</td>
      <td>1143</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/fluid_properties)

### Functional expansion tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.05%</td>
      <td>83.06%</td>
      <td>+0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>926</td>
      <td>927</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>189</td>
      <td>189</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/functional_expansion_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/functional_expansion_tools)

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.93%</td>
      <td>89.95%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3590</td>
      <td>3590</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>402</td>
      <td>401</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/heat_conduction)

### Level set

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.37%</td>
      <td>88.46%</td>
      <td>+0.10%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>319</td>
      <td>322</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>42</td>
      <td>42</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/level_set)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/level_set)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.32%</td>
      <td>86.43%</td>
      <td>+0.11%</td>
      <td>77.78%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10782</td>
      <td>10784</td>
      <td>+2</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1709</td>
      <td>1693</td>
      <td>-16</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/navier_stokes)

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>87.41%</td>
      <td>87.42%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1000</td>
      <td>1001</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>144</td>
      <td>144</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/optimization)

### Peridynamics

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>79.02%</td>
      <td>79.04%</td>
      <td>+0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3032</td>
      <td>3032</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>805</td>
      <td>804</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/peridynamics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/peridynamics)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.98%</td>
      <td>90.01%</td>
      <td>+0.03%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12516</td>
      <td>12517</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1393</td>
      <td>1389</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/phase_field)

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.24%</td>
      <td>96.33%</td>
      <td>+0.10%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10535</td>
      <td>10536</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>412</td>
      <td>401</td>
      <td>-11</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/porous_flow)

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>96.04%</td>
      <td>96.04%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4414</td>
      <td>4415</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>182</td>
      <td>182</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/reactor)

### Richards

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>97.49%</td>
      <td>97.49%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3339</td>
      <td>3342</td>
      <td>+3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>86</td>
      <td>86</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/richards)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/richards)

### Solid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.10%</td>
      <td>84.40%</td>
      <td>+0.30%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>238</td>
      <td>238</td>
      <td>-</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>45</td>
      <td>44</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/solid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/solid_properties)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.29%</td>
      <td>91.29%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6243</td>
      <td>6244</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>596</td>
      <td>596</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/stochastic_tools)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.45%</td>
      <td>89.55%</td>
      <td>+0.11%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26468</td>
      <td>26469</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3123</td>
      <td>3088</td>
      <td>-35</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/tensor_mechanics)

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>59f169</th>
      <th colspan=4>#23328 16234d</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.25%</td>
      <td>91.26%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12446</td>
      <td>12453</td>
      <td>+7</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1194</td>
      <td>1193</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23328/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23328/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 71.07% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 77.78% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad committed:
Unit testing of new functors

lindsayad commented:
@dschwen @GiudGiud I think this is ready for review

GiudGiud commented on framework/include/problems/SubProblem.h:
```diff
@@ -1129,25 +1230,94 @@ SubProblem::addFunctor(const std::string & name,
+          mooseError("We are requesting a non-AD functor from an AD object, but the true functor "
+                     "is AD. This means we could be dropping important derivatives. We will not "
+                     "allow this");
+        request_info_it = functor_to_request_info.erase(request_info_it);
```
we're deleting the request because we've filled it and we dont care anymore?
maybe add a comment?

GiudGiud commented on framework/include/problems/SubProblem.h:
```diff
@@ -1129,25 +1230,94 @@ SubProblem::addFunctor(const std::string & name,
+      std::get<0>(it->second) =
+          added_functor_is_ad ? SubProblem::TrueFunctorIs::AD : SubProblem::TrueFunctorIs::NONAD;
       existing_wrapper->assign(functor);
+      if constexpr (added_functor_is_ad)
```
add a comment that we re adding the other ADtype one in those two clauses

GiudGiud commented on framework/include/problems/SubProblem.h:
```diff
@@ -1129,25 +1230,94 @@ SubProblem::addFunctor(const std::string & name,
 
-  auto new_wrapper = std::make_unique<Moose::Functor<T>>(functor);
-  _functors[tid].emplace(std::make_pair("wraps_" + name, std::move(new_wrapper)));
+  // We are a new functor
```
```suggestion
  // We are a new functor, create the opposite ADType one and store it with other functors
```

GiudGiud commented on framework/src/problems/SubProblem.C:
```diff
@@ -1073,6 +1075,13 @@ void
+    for (auto & pr : map)
+    {
+      std::get<1>(pr.second)->timestepSetup();
+      std::get<2>(pr.second)->timestepSetup();
```
umm why do we need to do that?
shouldnt we be wrapping the same functor twice, AD and nonAD wrappers but still same object to set up?

GiudGiud reviewed (changes requested):
cleanup is pretty awesome

lindsayad commented on framework/include/base/MooseFunctor.h:
```diff
@@ -29,14 +29,26 @@
+/**
+ * Abstract base class that can be used to hold collections of functors
+ */
+class FunctorAbstract : public FaceArgInterface
```
because it's templated; you won't be able to store different instantiations of a template in the same container

lindsayad commented on framework/include/base/MooseFunctor.h:
```diff
@@ -138,7 +150,7 @@ class FunctorBase : public FaceArgInterface
   virtual bool isConstant() const { return false; }
 
-  bool hasFaceSide(const FaceInfo & fi, const bool fi_elem_side) const override;
+  virtual bool hasFaceSide(const FaceInfo & fi, const bool fi_elem_side) const override;
```
Haha. Yea up until very recently I did indeed remove `virtual` anyplace I found `override` ... but then when perusing doxygen pages I found it really helpful if I saw the `virtual` prefix in front for derived classes ... so I regret my earlier ways

lindsayad commented on framework/include/userobject/UserObject.h:
```diff
@@ -50,7 +49,6 @@ class UserObject : public MooseObject,
                    protected MeshChangedInterface,
                    protected ScalarCoupleable,
                    protected PerfGraphInterface,
-                   protected FunctorInterface,
```
`ResidualUO`?

lindsayad commented on framework/include/utils/ADifyFunctor.h:
```diff
@@ -0,0 +1,119 @@
+  using typename Moose::FunctorBase<T>::DotType;
+
+  ADifyFunctor(const FunctorBase<typename MetaPhysicL::RawType<T>::value_type> & non_ad_functor)
+    : FunctorBase<T>(non_ad_functor.functorName() + "_ad_ified", {EXEC_ALWAYS}),
```
I like `ad_wrapper` and I also think I like `ADWrapperFunctor` better than `ADifyFunctor`

lindsayad committed:
Address Guillaume review

- ADifyFunctor -> ADWrapperFunctor
- Doxygen
- Clarifying comments
- Only need to setup the true functor

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Only need to setup functors we own.

Other functors like variables and functions are owned by their
warehouses and have their respective setup calls in other places
of the code

lindsayad committed:
Don't allow functor envelopes to own true functors

This allows us to not worry about performing setup calls through
envelopes

lindsayad commented:
rebased to resolve merge conflict and tacked on one more commit after your approval @GiudGiud

GiudGiud reviewed (approved):
new changes look good.

moosebuild commented:
Job [Docker](https://civet.inl.gov/job/1386447/) on d6dfe87 : invalidated by @lindsayad

lindsayad commented:
thanks for the quick reviews!
